torch_geometric/nn/models/label_usage.py

```diff
@@ -0,0 +1,108 @@
+import torch
+import torch.nn.functional as F
+from torch import Tensor
+
+from torch_geometric.typing import Adj
+
+
+class LabelUsage(torch.nn.Module):
+    r"""The label usage operator for semi-supervised node classification,
+    as introduced in `"Bag of Tricks for Node Classification"
+    <https://arxiv.org/abs/2103.13355>`_ paper.
+
+    .. note::
+
+        When using the :class:`LabelUsage`, adjust the model's input dimension
+        accordingly to include both features and classes.
+
+    Args:
+        base_model: An instance of the model that will do the
+            inner forward pass.
+        num_classes (int): Number of classes in dataset
+        split_ratio (float): Proportion of true labels to use as features
+            during training (default: :obj:'0.5')
+        num_recycling_iterations (int): Number of iterations for the
+            label reuse procedure to cycle predicted soft labels
+            (default: :obj:'0')
+        return_tuple (bool): If true, returns (output, train_label_idx,
+            train_pred_idx) otherwise returns prediction output
+            (default :obj:'False')
+    """
+    def __init__(
+        self,
+        base_model: torch.nn.Module,
+        num_classes: int,
+        split_ratio: float = 0.5,
+        num_recycling_iterations: int = 0,
+        return_tuple: bool = False,
+    ):
+
+        super().__init__()
+        self.base_model = base_model
+        self.num_classes = num_classes
+        self.split_ratio = split_ratio
+        self.num_recycling_iterations = num_recycling_iterations
+        self.return_tuple = return_tuple
+
+    def forward(
+        self,
+        feat: Tensor,
+        edge_index: Adj,
+        y: Tensor,
+        mask: Tensor,
+        val_idx: Tensor,
+        test_idx: Tensor,
+    ):
+        r"""Forward pass using label usage algorithm.
+
+        Args:
+            feat (torch.Tensor): Node feature tensor of dimension (N,F)
+                where N is the number of nodes and
+                F is the number of features per node
+            edge_index (torch.Tensor or SparseTensor): The edge indices with
+                dimension (N,N) for a dense adjacency matrix or (2,E) for a
+                sparse adjacency representation where E is the number of edges
+            y (torch.Tensor): Node ground-truth labels tensor of dimension
+                of (N,) for 1D tensor or (N,1) for 2D tensor
+            mask (torch.Tensor): Global indices of all nodes in training
+                set or mini-batch with dimension (N_train,) where N_train
+                is the number of nodes in training set or mini-batch
+            val_idx (torch.Tensor): Node indices in validation set with
```

Comment:
the split should be done by the user , we either encapsulate the logic into a single variable `unlabeled_idx` or in the `mask` tensor IMO

Reply:
I agree user shouldn't have to explicitly pass their validation and test index, I think we only need their train index.  We can then infer the non-train index by computing the complement (we don't need to differentiate validation and test).  

Am I overlooking a reason why my earlier suggestion doesn't work for getting all unlabeled indices? 
```
unlabeled_mask = torch.ones(x.size(0), dtype=torch.bool)
unlabeled_mask[train_labels_idx] = False
```